2020/dec16/rules/rule.go

```diff
@@ -0,0 +1,64 @@
+package rules
+
+// NewRule is an ugly way to create a rule
+func NewRule(min1, max1, min2, max2 int, isDeparture bool) (Rule, error) {
+	// todo validate min < max
+	return Rule{
+		ranges: []ruleRange{
+			{min1, max1},
+			{min2, max2},
+		},
+		isDeparture: isDeparture,
+		lock: -1,
+	}, nil
+}
+
+// Rule defines one rule
+type Rule struct {
+	ranges      []ruleRange
+	positions   int
+	maxMask     int
+	isDeparture bool
+	lock        int
+}
+
+// IsValid returns whether the value matches the rule
+func (ru *Rule) IsValid(v int) bool {
+	for _, rg := range ru.ranges {
+		if rg.isValid(v) {
+			return true
+		}
+	}
+	return false
+}
+
+func (ru *Rule) MatchTicket(ticket Ticket) {
+	for i, v := range ticket {
+		if !ru.IsValid(v) {
+			ru.positions &= (1 << i) ^ ru.maxMask
+		}
+	}
+}
+
+func (ru *Rule) TryLock() (int, bool) {
```

Comment:
_If you prick us, do we not bleed? if you tickle us, do we not laugh? if you poison us, do we not die?_